### ISSUE TITLE
Using the plugin with root user

### ISSUE THREAD

**Issue by mikomarrache** [donapieppo/libnss-ato#14]
I'm not able to use the plugin with the root user but it works with another user.

Here is the config I'm using:

root:x:0:0:Root:/root:/bin/bash

Here are the logs I see (/var/log/secure):

sshd[2235]: Accepted password for myuser from 172.20.0.1 port 46328 ssh2
sshd[2235]: pam_systemd(sshd:session): Failed to create session: No such file or directory
sshd[2235]: pam_unix(sshd:session): session opened for user myuser by (uid=0)
sshd[2235]: fatal: login_get_lastlog: Cannot find account for uid 500
sshd[2235]: pam_unix(sshd:session): session closed for user myuser
sshd[2235]: syslogin_perform_logout: logout() returned an error
sshd[2238]: fatal: mm_request_send: write: Broken pipe

Am I missing something?

**Comment by vmspike** [comment]
See libnss_ato.c: 
```
/* for security reasons */
#define MIN_UID_NUMBER   500
#define MIN_GID_NUMBER   500
```
So if you want to use it with root, you have to decrease that values to 0 as I understand.

**Comment by mikomarrache** [comment]
I tried that but I got side effects. At least the mysql user disappeared for some reason. It looks like the plugin is used for more than just SSH/login.

**Comment by donapieppo** [comment]
I tried to comment in libnss_ato.c the lines that checks MIN and MAX uid number

```
        /*
        if ( conf->pw_uid < MIN_UID_NUMBER )
                conf->pw_uid = MIN_UID_NUMBER;

        if ( conf->pw_gid < MIN_GID_NUMBER )
                conf->pw_gid = MIN_GID_NUMBER;
        */
```
and the command id works:
```
# id pippo
uid=0(root) gid=0(root) groups=0(root)
```
Can you describe what doesn't work for you so I can replicate your system?

**Comment by mikomarrache** [comment]
Yes I also tried and it works but then the plugin becomes problematic since
other applications that depend on NSS would get the wrong user. Just as an
example, installing mysql after the plugin is configured is problematic
since the mysql user (27) will not be created. One of the forks added some
checks so that the plugin is used only for SSH.

Le jeu. 27 sept. 2018 à 15:01, Donapieppo <notifications@github.com> a
écrit :

**Comment by donapieppo** [comment]
I think that every system operation like, for example, creating new users (I think that mysql package checks the existence of mysql user and, as MIN_UID_NUMBER = 0, it finds it and does not create it)
has that problem.

I suggest that you create a file like `/etc/nsswitch.conf.ato` (with passwd: files ato)
and `/etc/nsswitch.conf.noato` (with passwd: files) 

Then to install packages:
```
cp /etc/nsswitch.conf.noato /etc/nsswitch.conf
apt-get install  mysql-server
cp /etc/nsswitch.conf.ato /etc/nsswitch.conf
```

**Comment by vmspike** [comment]
This issue is already referred in #11.
For APT it can be better to create file `/etc/apt/apt.conf.d/00libnss-ato` with the content like
```
DPkg::Pre-Invoke  {"/bin/ln -sf /etc/nsswitch.conf.noato /etc/nsswitch.conf";};
DPkg::Post-Invoke {"/bin/ln -sf /etc/nsswitch.conf.ato   /etc/nsswitch.conf";};
```
and separate files `/etc/nsswitch.conf.noato` and `/etc/nsswitch.conf.ato`, to temporary disable `libnss-ato` for packages installation.
If it's a way to limit `libnss-ato` for use with ssh only would like to know it too.

**Comment by donapieppo** [comment]
Cool solution  (Pre- Post - Invoke) for apt!!!
